### ISSUE TITLE
Tuples don't work!

### ISSUE THREAD

**Issue by drozzy** [lorenzoh/DataLoaders.jl#12]
Julia Versions tried: 1.5.2, 1.5.3. Ubuntu. 20.04.

If I run the tutorial example I get this error (note that it works if I just pass `x` but not if I pass the tuple `(x,y)`):
```
using DataLoaders

x = rand(128, 10000)  #  10000 observations of size 128
y = rand(1, 10000)

dataloader = DataLoaders.DataLoader((x, y), 16)

for (xs, ys) in dataloader
    @assert size(xs) == (128, 16)
    @assert size(ys) == (1, 16)
end
```

Error:
```
ERROR: MethodError: no method matching iterate(::MLDataPattern.DataSubset{DataLoaders.BatchViewCollated{Tuple{Array{Float64,2},Array{Float64,2}}},Int64,LearnBase.ObsDim.Undefined})
Closest candidates are:
  iterate(!Matched::LibGit2.GitRebase) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LibGit2/src/rebase.jl:48
  iterate(!Matched::LibGit2.GitRebase, !Matched::Any) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/LibGit2/src/rebase.jl:48
  iterate(!Matched::ThreadPools.ResultIterator) at /home/andriy/.julia/packages/ThreadPools/hPQNy/src/qpool.jl:172
  ...
Stacktrace:

_zip_iterate_some at ./iterators.jl:352

_zip_iterate_some at ./iterators.jl:354

_zip_iterate_all at ./iterators.jl:344

iterate at ./iterators.jl:334

iterate at ./generator.jl:44

collect(::Base.Generator{Base.Iterators.Zip{Tuple{Tuple{Array{Float64,2},Array{Float64,2}},MLDataPattern.DataSubset{DataLoaders.BatchViewCollated{Tuple{Array{Float64,2},Array{Float64,2}}},Int64,LearnBase.ObsDim.Undefined}}},Base.var"#3#4"{typeof(LearnBase.getobs!)}}) at ./array.jl:686

map at ./abstractarray.jl:2248

iterate(::MLDataPattern.BufferGetObs{Tuple{Array{Float64,2},Array{Float64,2}},MLDataPattern.ObsView{MLDataPattern.DataSubset{DataLoaders.BatchViewCollated{Tuple{Array{Float64,2},Array{Float64,2}}},Int64,LearnBase.ObsDim.Undefined},DataLoaders.BatchViewCollated{Tuple{Array{Float64,2},Array{Float64,2}}},LearnBase.ObsDim.Undefined}}) at /home/andriy/.julia/packages/MLDataPattern/KlSmO/src/dataiterator.jl:524

top-level scope at /home/andriy/Projects/Projectjl/src/02.dataset.jl:17

include_string(::Function, ::Module, ::String, ::String) at ./loading.jl:1091

invokelatest(::Any, ::Any, ::Vararg{Any,N} where N; kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at ./essentials.jl:710

invokelatest(::Any, ::Any, ::Vararg{Any,N} where N) at ./essentials.jl:709

inlineeval(::Module, ::String, ::Int64, ::Int64, ::String; softscope::Bool) at /home/andriy/.vscode/extensions/julialang.language-julia-1.0.10/scripts/packages/VSCodeServer/src/eval.jl:185

(::VSCodeServer.var"#61#65"{String,Int64,Int64,String,Module,Bool,VSCodeServer.ReplRunCodeRequestParams})() at /home/andriy/.vscode/extensions/julialang.language-julia-1.0.10/scripts/packages/VSCodeServer/src/eval.jl:144

withpath(::VSCodeServer.var"#61#65"{String,Int64,Int64,String,Module,Bool,VSCodeServer.ReplRunCodeRequestParams}, ::String) at /home/andriy/.vscode/extensions/julialang.language-julia-1.0.10/scripts/packages/VSCodeServer/src/repl.jl:124

(::VSCodeServer.var"#60#64"{String,Int64,Int64,String,Module,Bool,Bool,VSCodeServer.ReplRunCodeRequestParams})() at /home/andriy/.vscode/extensions/julialang.language-julia-1.0.10/scripts/packages/VSCodeServer/src/eval.jl:142

hideprompt(::VSCodeServer.var"#60#64"{String,Int64,Int64,String,Module,Bool,Bool,VSCodeServer.ReplRunCodeRequestParams}) at /home/andriy/.vscode/extensions/julialang.language-julia-1.0.10/scripts/packages/VSCodeServer/src/repl.jl:36

(::VSCodeServer.var"#59#63"{String,Int64,Int64,String,Module,Bool,Bool,VSCodeServer.ReplRunCodeRequestParams})() at /home/andriy/.vscode/extensions/julialang.language-julia-1.0.10/scripts/packages/VSCodeServer/src/eval.jl:110

with_logstate(::Function, ::Any) at ./logging.jl:408

with_logger at ./logging.jl:514

(::VSCodeServer.var"#58#62"{VSCodeServer.ReplRunCodeRequestParams})() at /home/andriy/.vscode/extensions/julialang.language-julia-1.0.10/scripts/packages/VSCodeServer/src/eval.jl:109

#invokelatest#1 at ./essentials.jl:710

invokelatest(::Any) at ./essentials.jl:709

macro expansion at /home/andriy/.vscode/extensions/julialang.language-julia-1.0.10/scripts/packages/VSCodeServer/src/eval.jl:27

(::VSCodeServer.var"#56#57")() at ./task.jl:356
```

Output of `]st`:
```
 [6e4b80f9] BenchmarkTools v0.5.0
  [336ed68f] CSV v0.8.2
  [a93c6f00] DataFrames v0.21.8
  [2e981812] DataLoaders v0.1.1
  [743a1d0a] DataTables v0.1.0
  [aae7a2af] DiffEqFlux v1.25.0
  [587475ba] Flux v0.11.2
  [7f8f8fb0] LearnBase v0.3.0
  [9920b226] MLDataPattern v0.5.4
  [cc2ba9b6] MLDataUtils v0.5.2
  [14b8a8f1] PkgTemplates v0.7.13
  [3cdcf5f2] RecipesBase v1.1.1
  [f3b207a7] StatsPlots v0.14.17
  [09ab397b] StructArrays v0.4.4
  [b8865327] UnicodePlots v1.3.0
```

**Comment by drozzy** [comment]
Funny thing is I'm sure it worked before...

**Comment by darsnack** [comment]
Good catch. Until this is fixed, you can work around it by wrapping your data in a new type that returns tuples for `getobs`. [Here](https://github.com/SciML/DiffEqFlux.jl/pull/387#issuecomment-734007664) is a recent example of that for MLDatasets.MNIST.

**Comment by drozzy** [comment]
Not sure I understand that example (still new to Julia) :-)
But I just ended converting my views to full-blown matrices and using Flux.DataLoader :-)

**Comment by lorenzoh** [comment]
Hey @drozzy, the issue was only on Julia sessions started without multiple threads, so you should not get the error if you start with julia using `julia -t n` where `n` is the number of threads to use. This is encouraged since only then will you be able to benefit from the parallelism.

I've opened a PR (#13) that addresses this issue and allows using `DataLoader`s that only use the main thread which will be the default behavior if `Threads.nthreads() == 1`.

Starting julia with the `-t` option should definitely fix the issue for now, and once the PR is merged I'll release a patch version so the single-threaded version works as well.

**Comment by drozzy** [comment]
Hi, @lorenzoh, thanks but I'm just using VSCode and whatever Julia it is starting by default. I mostly run Julia code inline (with option+enter).

**Comment by lorenzoh** [comment]
You can change the number of threads Julia starts with in VSCode with the "Julia: Num Threads" setting which you can find in the settings UI. Let me know if that works :)

**Comment by drozzy** [comment]
The thing is I'm using Remote-SSH with VSCode that starts Julia process on another computer, so I'm not sure my local VSCode settings will work remotely.
Anyways, if the new PR fixes this - I can wait. :-)

**Comment by lorenzoh** [comment]
In the settings page there should be tabs at the tabs for where you want to change the setting and there should be a Remote option.

**Comment by drozzy** [comment]
Oh wow, thank you!
So should I always start it with multiple threads? Seems like a good idea to do regardless!

**Comment by lorenzoh** [comment]
I would suggest that, especially if you're running larger data pipelines. You can set the number of threads equal to the number of physical cores your machine has.

**Comment by drozzy** [comment]
Aha! I can confirm that DataLoaders.jl works if I start Julia with 8 threads!

**Comment by lorenzoh** [comment]
Good to hear :)

Since https://github.com/JuliaRegistries/General/pull/25481 is merged, you should also be able to install the new version using `]update DataLoaders`.
